### ISSUE TITLE
Passing userdata pointer back to custom allocator functions

### ISSUE THREAD

**Issue by gyrovorbis** [tidwall/hashmap.c#15]
Yooo! Thank you so much for this neat little project. I've been writing a bunch of C code for a common runtime environment for multi-language interop, and I finally reached a point where I can't avoid trying to not have to deal with Hash Map data types anymore.... ANYWAY, I love how elegant, compact, and well-written this is. It actually made it really easy for me to "subtype" and encapsulate it within my own framework...

My ONLY complaint is about the custom allocator support, which I'm extremely grateful for. The only issue is that I have multilevel stateful allocators everywhere, so I can't really hook into them using stateless allocation functions. 

Is there any way you could pass the void* ud state back to the allocators? Here's my use-case:

typedef struct GblHashMap {
    struct hashmap* pImpl_;
    GblContext hCtx;
} GblHashMap;

I'm basically C "inheriting" from your opaque type, and am just passing the pointer to the outer GblHashMap object as the ud. The "GblContext" object is what contains the custom allocators, error loggers, etc.

GBL_INLINE GBL_RESULT gblHashMapSet(GblHashMap* pMap, void* pItem, void** ppData) {
    GBL_API_ASSERT(pMap);
    GBL_API_BEGIN(pMap->hCtx);
    GBL_API_VERIFY_POINTER(pItem);
    GBL_API_VERIFY_POINTER(ppData);
    *ppData = hashmap_set(pMap->pImpl, pItem);
    GBL_API_END();
}

GBL_INLINE GBL_RESULT gblHashMapDelete(GblHashMap* pMap, void* pItem, void** ppData) {
    GBL_API_ASSERT(pMap);
    GBL_API_BEGIN(pMap->hCtx);
    GBL_API_VERIFY_POINTER(pItem);
    GBL_API_VERIFY_POINTER(ppData);
    *ppData = hashmap_delete(pMap->pImpl, pItem);
    GBL_API_END();
}

Here is me with the custom allocator callbacks, but you can see I don't have access to the ud pointer or any external state... which means my macros are just falling back to the default C standard functions.

GBL_INLINE void* gblHashAlloc_(size_t bytes) {
    GBL_API_BEGIN(NULL);
    return GBL_API_MALLOC(bytes);
}

GBL_INLINE void* gblHashRealloc_(void* pData, size_t bytes) {
    GBL_API_BEGIN(NULL);
    return GBL_API_REALLOC(pData, bytes);
}

GBL_INLINE void gblHashFree_(void* pData) {
    GBL_API_BEGIN(NULL);
    GBL_API_FREE(pData);
}

**Comment by gyrovorbis** [comment]
Sounds good and no problem. I've just modified it locally. By the way, the read performance is fantastic. I am writing a low-level C/++ framework that offers unified data structures in both C and C++ which can be driven from either language. I based the core implementation in C on yours with a slightly more STL-ish API, called it HashSet. I then implemented a templated C++ wrapper around it called HashSet and a HashMap above that with an exact std::unordered_map API. 

Anyway, long story short, even with function pointers/dynamic binding for the hash and comparison functions (which templates in STL don't have to have), the read function almost exactly as fast and slightly better than std::unordered_map and Qt's QHashMap built around your algorithm! The write performance is slightly slower but is still competitive, and this is with custom allocators in C and none in C++!
